### PR TITLE
Add MCMC diagnostics to output of `marginals`

### LINKED ISSUE #8: Report effective sample size, etc.

### PR DESCRIPTION
**Pull request by null-a**
This PR extends the output of the marginals helper with MCMC diagnostic information. (Closes #8.) For example:

```python
import pandas as pd
from brmp import defm
from brmp.fit import marginals
df = pd.DataFrame({'y': [0, 0.1, 1.0]})
fit = defm('y~1', df).fit()
print(marginals(fit))
```

Produces:

```
             mean    sd  2.5%   25%   50%   75% 97.5%  n_eff r_hat
b_intercept  0.15  0.35 -0.31 -0.07  0.05  0.39  0.79  24.18  0.90
      sigma  0.66  0.21  0.43  0.51  0.59  0.74  1.07   6.82  1.44
```

There are a couple of things I'm unsure about here.

This first is whether my conditional use of `gelman_rubin` or `split_gelman_rubin` (depending on the number of chains run) is what we want?

The other is whether this is computing the right thing for `n_eff`. I ran a simple example multiple times with this and brms to sanity check the output, and it's not clear to me both sets of results are similarly distributed. See details below.

<details>

Intercept:

| brms | brmp  |
|---|---|
| 1673  | 1183   |
|1143 | 1123  |
|1428 | 1029    |
|1272 | 930  |
|1596 | 1372  |

Sigma:

| brms | brmp  |
|---|---|
| 1375 |  1094 |
| 1306  | 1201 |
|  1364 | 1176 |
| 1461 | 1087 |
| 1522 | 1208 |

```python
import pandas as pd
from brmp import defm
from brmp.fit import marginals
from brmp.priors import Prior
from brmp.family import Normal, HalfNormal
from brmp.numpyro_backend import backend as numpyro

df = pd.DataFrame({'y': [0, 0.1, 1.0]})
model = defm(
    'y~1',
    df,
    priors=[
        Prior(('b',), Normal(0., 1.)),
        Prior(('resp', 'sigma',), HalfNormal(1.))
    ])
fit = model.fit(num_chains=4, iter=1000, warmup=1000, backend=numpyro)
print(marginals(fit))
```

```R
library(brms)
df = data.frame(y=c(0.,0.1,1.))
fit = brm(bf(y~1, center=0), data=df, prior=c(set_prior(class='b', 'normal(0, 1)'), set_prior(class='sigma', "normal(0, 1)")))
fit$fit
```

I think these describe the same model. In brms, I believe that the normal prior gets turned into a half normal automatically.

</details>

### CODE REVIEW

File: brmp/fit.py [hunk 1]
```diff
@@ -139,6 +141,28 @@ def summary(arr, qs=default_quantiles, row_labels=None):
...
+        return None # Too few samples.
+    else:
+        return diags.effective_sample_size(samples)
+
+def compute_diag_or_default(diag, samples):
+    val = diag(samples)
+    if val is not None:
+        return val
+    else:
+        return np.zeros((samples.shape[2],))
```
Comment (neerajprad) [inline brmp/fit.py]: Do you think 0 is a good default value, or should we default to a NaN array instead?

**Comment by neerajprad** [comment]
LGTM. Regarding:

> This first is whether my conditional use of gelman_rubin or split_gelman_rubin (depending on the number of chains run) is what we want?

I think that's a reasonable choice. I am not fully sure if that's what brms or Stan does with a single chain.

> The other is whether this is computing the right thing for n_eff.

It seems fishy that we are consistently under-approximating the effective sample size. This is worth looking into for sure separately. cc. @fehiepsi - have you observed this with Stan previously?

**Comment by fehiepsi** [comment]
For most models I have played with, NumPyro will give comparable or better n_eff than Stan (NumPyro does not suffer from [this issue of Stan](https://discourse.mc-stan.org/t/nuts-misses-u-turns-runs-in-circles-until-max-treedepth/9727/6), Stan devs fixed that issue recently but the fix has some side effects for other types of models). However, Stan reports n_eff including warmup samples, while NumPyro does not include those warmup samples. So the reported n_eff from Stan will usually 1.5-2x greater than NumPyro if num_samples=num_warmup. E.g.in brms result, it reports `n_eff = 1673` while `iter=1000`.

This difference has been a wrong understanding about NUTS in Pyro for a while. And we have put no attempt to clarify it (my bad). In this [blog post](https://jeffpollock9.github.io/going-NUTS-with-pyro-and-pystan/), quoted from the author
_It also looks like the sampling algorithm in stan was a bit more efficient as the estimated sample size (n_eff) is much bigger in stan (around 5,000) compared to pyro (around 3,000) for 3,000 post warmup draws. This is assuming that the n_eff calculation is the same in stan and pyro._

@neerajprad We have no blog post/docs to clarify that difference. I think that we should do it now.

@null-a Could you try to add `collect_warmup=True` to `mcmc.run(...)` to collect all warmup samples and print out result?

**Comment by neerajprad** [comment]
> However, Stan reports n_eff including warmup samples, while NumPyro does not include those warmup samples. So the reported n_eff from Stan will usually 1.5-2x greater than NumPyro if num_samples=num_warmup.

That's reassuring to know, and explains why we are seeing lower n_eff values! 

> @neerajprad We have no blog post/docs to clarify that difference. I think that we should do it now, e.g. mentioned about this in the docs. WDYT?

I agree we should highlight this in a `note` in both Pyro and NumPyro. Users often compare with Stan, and might unnecessarily be disappointed or concerned. We should also comment on that blog post. :) Other users are likely to bump into posts like that.

**Comment by fehiepsi** [comment]
I just test with the latest version of PyStan and see that PyStan also reported n_eff for post-warmup samples. My previous post is completely wrong. Sorry about that! I will investigate why we get a worse result than Stan.

**Comment by fehiepsi** [comment]
Updated: The way Stan computes n_eff is different from Pyro. I made this [gist](https://nbviewer.jupyter.org/gist/fehiepsi/f03eeaa85b7d00a4d6d8214fea96606d) to illustrate that. First, I used Stan to draw 1000 samples, we can see that Stan reported n_eff=3903 for `beta[1]`, which is unreasonable to me. Then I used NumPyro summary to report the result for these samples: I get `n_eff = 941.15` for `beta[1]`. Finally, I checked the result with arviz and got a similar result to NumPyro: `n_eff = 944.0`. I have no idea why Stan reports such high n_eff (based on the result, I have mistakenly thought that Stan reports n_eff including warmup samples, but it turns out that the difference comes from the way Stan computes n_eff).

**Comment by neerajprad** [comment]
Let us move the discussion to numpyro, where it belongs. That shouldn't block this PR.

File: brmp/fit.py (hunk 1 above)
Comment (null-a) [reply brmp/fit.py]: No, good point, probably not. I've pushed a commit that switches this to nan, though I'm not sure that it improves things much. (When I first saw this in numpyro I though it was a bug, IIRC.) I think this is OK for now though, and I've opened #28 on the possibility of doing something else eventually.

**Comment by null-a** [comment]
> I think that's a reasonable choice.

Good to know, thanks.

I've switched the default for missing diag. info to nan as suggested, and the `n_eff` issue is picked up elsewhere so I'll go ahead and merge.

Thanks!

**Pull request merged**